lib/galaxy/config/__init__.py

```diff
@@ -861,6 +872,22 @@ def _process_config(self, kwargs):
                 'filters': ['stack']
             }
 
+    def _configure_dataset_storage(self):
+        # The default for `file_path` has changed in 20.05; we may need to fall back to the old default
+        self._set_alt_paths('file_path', self._in_data_dir('files'))  # this is called BEFORE guessing id/uuid
+        ID, UUID = 'id', 'uuid'
+        if self.is_set('object_store_store_by'):
+            assert self.object_store_store_by in [ID, UUID], "Invalid value for object_store_store_by [{}]".format(self.object_store_store_by)
+        else:
+            self.object_store_store_by = ID
+            if not self.is_set('file_path'):  # choose based on default dir name
+                if os.path.basename(self.file_path) == 'objects':
+                    self.object_store_store_by = UUID
+            else:  # choose based on dataset filename format
+                filename = find_dataset(self.file_path)
+                if not filename or is_uuid(filename):  # dir has no datasets, or contains a uuid dataset
+                    self.object_store_store_by = UUID
```

Comment:
Can we resolve

> The reason for the above error is that we assume that if `file_path` is set, we must be storing by id - which is not always correct.

with:

```suggestion
        elif os.path.basename(self.file_path) == 'objects':
            self.object_store_store_by = UUID
        else:
            self.object_store_store_by = ID
```

Reply:
OK. We'll just document the assumptions about `objects` in the `file_path` description, as per @mvdbeek .  I'll update the docs. Committing this now; I may be able to simplify the conditional logic here in a min..

@mvdbeek : I'm switching the check for `files` before `objects` in a separate commit.